client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentEventReaderImpl.java

```diff
@@ -0,0 +1,380 @@
+        readState.cancel();
+        asyncInput.close();
+    }
+
+    @Override
+    public boolean isClosed() {
+        return asyncInput.isClosed();
+    }
+
+    // region Cursor
+
+    @Override
+    @Synchronized
+    public void setOffset(long offset) {
+        log.trace("SetOffset {}", offset);
+        Preconditions.checkArgument(offset >= 0);
+        Exceptions.checkNotClosed(asyncInput.isClosed(), this);
+        if (offset != this.cursorOffset) {
+            readState.cancel();
+            this.cursorOffset = offset;
+        }
+    }
+
+    @Override
+    @Synchronized
+    public long getOffset() {
+        return cursorOffset;
+    }
+
+    @Override
+    @Synchronized
+    public void setEndOffset(long offset) {
+        log.trace("SetEndOffset {}", offset);
+        Preconditions.checkArgument(offset >= 0);
+        Exceptions.checkNotClosed(asyncInput.isClosed(), this);
+        this.endOffset = offset;
+    }
+
+    @Override
+    @Synchronized
+    public long getEndOffset() {
+        return endOffset;
+    }
+
+    @Override
+    @Synchronized
+    public CompletableFuture<ByteBuffer> readAsync() throws EndOfSegmentException {
+
+        // prepare a promise
+        if (cursorOffset >= this.endOffset) {
+            log.debug("All events up to the configured end offset:{} have been read", endOffset);
+            throw new EndOfSegmentException(END_OFFSET_REACHED);
+        }
+        CompletableFuture<ByteBuffer> promise = readState.reset(cursorOffset);
+        promise.whenComplete(this::readCompleted);
+
+        // kick off or continue processing the request
+        issueRequestIfNeeded();
+
+        return promise;
+    }
+
+    @Synchronized
+    private void readCompleted(ByteBuffer result, Throwable th) {
+        if (th != null) {
+            // an event could not be read; don't update the cursor position.
+            log.trace("Failed to read an event @ " + cursorOffset, th);
+
+            // cancel any outstanding request now to ensure that handleRequest doesn't affect the read state
+            if (outstandingRequest != null && !outstandingRequest.isDone()) {
+                log.trace("Cancel outstanding read request for segment {}", asyncInput.getSegmentId());
+                outstandingRequest.cancel(false);
+                outstandingRequest = null;
+            }
+        }
+        if (result != null) {
+            // an event was read successfully; update the cursor position.
+            cursorOffset = readState.offset;
+            log.trace("Read an event @ {}", cursorOffset);
+        }
+    }
+
+    @VisibleForTesting
+    ReadState getReadState() {
+        return readState;
+    }
+
+    static class ReadState {
+        private final ByteBuffer headerReadingBuffer = ByteBuffer.allocate(WireCommands.TYPE_PLUS_LENGTH_SIZE);
+        private CompletableFuture<ByteBuffer> outstandingPromise = CompletableFuture.completedFuture(null);
+        private ByteBuffer resultBuffer;
```

Comment:
It would be good if we didn't have to allocate a new buffer every single time. We have a class CircularBuffer for this purpose.

Reply:
Why is it copied? ByteBufs from netty should be direct references. If we are going to copy it it should be so that we can issue another request. Which we should so that we aren't network latency bound.